src/app/helpers/csv.js

```diff
@@ -0,0 +1,218 @@
+ * @param {Object} csvData
+ * @param {Array<String>} csvData.headers           Array of csv properties
+ * @param {Array<Array<String>>} csvData.contacts   Array of csv contacts
+ *
+ * @return {Array<Array<Object>>}                   pre-vCard contacts
+ *
+ * @dev  Some csv property may be assigned to several pre-vCard contacts,
+ *       so an array of new headers is returned together with the pre-vCard contacts
+ */
+const parse = ({ headers = [], contacts = [] }) => {
+    if (contacts.length === 0) {
+        return [];
+    }
+    const { headers: standardHeaders, contacts: standardContacts } = standarize(
+        getNonEmptyCsvData({ headers, contacts })
+    );
+    const translator = standardHeaders.map(toPreVcard);
+    return standardContacts
+        .map((contact) =>
+            contact
+                .map((header, i) => translator[i](header))
+                // some headers can be mapped to several properties, so we need to flatten
+                .reduce((acc, val) => acc.concat(val), [])
+        )
+        .map((contact) => contact.filter((preVcard) => !!preVcard));
+};
+
+/**
+ * Transform csv properties and csv contacts into pre-vCard contacts,
+ * re-arranging them in the process
+ * @param {Object} csvData
+ * @param {Array<String>} csvData.headers           Array of csv properties
+ * @param {Array<Array<String>>} csvData.contacts   Array of csv contacts
+ *
+ * @return {Array<Object>}                          Array of pre-vCard contacts
+ *
+ * @dev  headers are arranged as headers = [[group of headers to be combined in a vCard], ...]
+ *       preVcardContacts is an array of pre-vCard contacts, each of them containing pre-vCards
+ *       arranged in the same way as the headers:
+ *       preVcardContacts = [[[group of pre-vCard properties to be combined], ...], ...]
+ */
+export const prepare = ({ headers = [], contacts = [] }) => {
+    const preVcardContacts = parse({ headers, contacts });
+    if (!preVcardContacts.length) {
+        return [];
+    }
+
+    // detect csv properties to be combined in preVcardContacts and split header indices
+    const nonCombined = [];
+    const combined = preVcardContacts[0].reduce((acc, { combineInto, combineIndex: j }, i) => {
+        if (combineInto) {
+            if (!acc[combineInto]) {
+                acc[combineInto] = [];
+            }
+            acc[combineInto][j] = i;
+            // combined will look like e.g.
+            // { 'fn-main': [2, <empty item(s)>, 3, 5, 1], 'fn-yomi': [<empty item(s)>, 6, 7] }
+            return acc;
+        }
+        nonCombined.push(i);
+        return acc;
+    }, {});
+
+    for (const combination of Object.keys(combined)) {
+        // remove empty items from arrays in combined
+        combined[combination] = combined[combination].filter((n) => n !== null);
+    }
+
+    // Arrange pre-vCards respecting the original ordering outside header groups
+    const preparedPreVcardContacts = contacts.map((_) => []);
```

Comment:
Eslint feedback: `_` unused

Reply:
yeah, I thought it was fixed with `(_)`, but actually it's ok to simply have `()` here!